### PR TITLE
Remove namespace from charts and split them into `linkerd-crd` and `linkerd-control-plane`

### LINKED ISSUE #6584: Remove the namespace resource from the Helm charts
It is considered bad practice to have the charts create their namespace, specially since Helm v3. A few reasons for this:
- Just like you would do with `kubectl --namespace xxx apply`, it's more flexible to leave `helm install` instruct in a flag the namespace it'd like the resources to be deployed into.
- Now that Tiller is gone, the chart config is no longer stored in Tiller's namespace, but in the namespace passed to `helm install --namespace`. If nothing was passed, the config secret is stored in the `default` namespace regardless of the namespace used in the chart's templates. That secret can contain private data, so that's not secure.

Removing the namespace from the templates requires to take these points into account.
- Have the CLI to manually generate the yaml for the namespace.
- The namespace can no longer have this specific linkerd metadata:
  - The `linkerd.io/inject: disabled` annotation. No big deal, we'll just have to add that to each control plane workload.
  - The `config.linkerd.io/admission-webhooks: disabled` label. Without this the injector will get called for control plane pod restarts, but it shouldn't mutate the pod given the presence of the `linkerd.io/inject: disabled` annotation.
  - The `linkerd.io/is-control-plane: "true"` label. From what I could tell this is not relied on.
  - The `linkerd.io/control-plane-ns` label. Used for identifying all the resources belonging to linkerd, and used for `kubectl apply --prune`. Since the namespace isn't touched during upgrades, it seems this doesn't matter. Integrations tests though will need to manually remove namespaces for cleanup.
  - Linkerd extensions rely on the `linkerd.io/extension` label. If we can't find an easy workaround for this, we could add a simple post-install hook for just adding that label.
- For folks having installed without using `helm install --namespace`, a simple upgrade to this new way won't likely be possible. Will have to figure if it's possible to do it without downtime, and document the procedure. Maybe we can consider still leaving the namespace template in the chart, but disabled by default, and just instruct to enable it in the upgrade docs for folks not wanting to migrate to the new way.

Ref #6463 #3211

### PR DESCRIPTION
**Pull request by alpeb**
Fixes #6584 #6620 #7405

# Namespace Removal

With this change, the `namespace.yaml` template is rendered only for CLI installs and not Helm, and likewise the `namespace:` entry in the namespace-level objects (using a new `partials.namespace` helper).

The `installNamespace` and `namespace` entries in `values.yaml` have been removed.

There in the templates where the namespace is required, we moved from `.Values.namespace` to `.Release.Namespace` which is filled-in automatically by Helm. For the CLI, `install.go` now explicitly defines the contents of the `Release` map alongside `Values`.

The proxy-injector has a new `linkerd-namespace` argument given the namespace is no longer persisted in the `linkerd-config` ConfigMap, so it has to be passed in. To pass it further down to `injector.Inject()` without modifying the `Handler` signature, a closure was used.

------------
Update: Merged-in #6638: Similar changes for the `linkerd-viz` chart:

Stop rendering `namespace.yaml` in the `linkerd-viz` chart.

The additional change here is the addition of the `namespace-metadata.yaml` template (and its RBAC), _not_ rendered in CLI installs, which is a Helm `post-install` hook, consisting on a Job that executes a script adding the required annotations and labels to the viz namespace using a PATCH request against kube-api. The script first checks if the namespace doesn't already have an annotations/labels entries, in which case it has to add extra ops in that patch.

---------
Update: Merged-in the approved #6643, #6665 and #6669 which address the `linkerd2-cni`, `linkerd-multicluster` and `linkerd-jaeger` charts. 

Additional changes from what's already mentioned above:
- Removes the install-namespace option from `linkerd install-cni`, which isn't found in `linkerd install` nor `linkerd viz install` anyways, and it would add some complexity to support.
- Added a dependency on the `partials` chart to the `linkerd-multicluster-link` chart, so that we can tap on the `partials.namespace` helper.
- We don't have any more the restriction on having the muticluster objects live in a separate namespace than linkerd. It's still good practice, and that's the default for the CLI install, but I removed that validation.

Finally, as a side-effect, the `linkerd mc allow` subcommand was fixed; it has been broken for a while apparently:

```console
$ linkerd mc allow --service-account-name foobar
Error: template: linkerd-multicluster/templates/remote-access-service-mirror-rbac.yaml:16:7: executing "linkerd-multicluster/templates/remote-access-service-mirror-rbac.yaml" at <include "partials.annotations.created-by" $>: error calling include: template: no template "partials.annotations.created-by" associated with template "gotpl"
```
---------
Update: see helm/helm#5465 describing the current best-practice

# Core Helm Charts Split

This removes the `linkerd2` chart, and replaces it with the `linkerd-crds` and `linkerd-control-plane` charts. Note that the viz and other extension charts are not concerned by this change.

Also note the original `values.yaml` file has been split into both charts accordingly.

### UX

```console
$ helm install linkerd-crds --namespace linkerd --create-namespace linkerd/linkerd-crds
...
# certs.yaml should contain identityTrustAnchorsPEM and the identity issuer values
$ helm install linkerd-control-plane --namespace linkerd -f certs.yaml linkerd/linkerd-control-plane
```

### Upgrade

As explained in #6635, this is a breaking change. Users will have to uninstall the `linkerd2` chart and install these two, and eventually rollout the proxies (they should continue to work during the transition anyway).

### CLI

The CLI install/upgrade code was updated to be able to pick the templates from these new charts, but the CLI UX remains identical as before.

### Other changes

- The `linkerd-crds` and `linkerd-control-plane` charts now carry a version scheme independent of linkerd's own versioning, as explained in #7405.
- These charts are Helm v3, which is reflected in the `Chart.yaml` entries and in the removal of the `requirements.yaml` files.
- In the integration tests, replaced the `helm-chart` arg with `helm-charts` containing the path `./charts`, used to build the paths for both charts.

### Followups

- Now it's possible to add a `ServiceProfile` instance for Destination in the `linkerd-control-plane` chart.

### CODE REVIEW

**Review by adleong — APPROVED** [review]
Nice.  What will the upgrade procedure be for Helm users? I take it they have to do a uninstall and then reinstall of the helm chart?  Assuming they use the same issuer certificates, can they keep the data plane running the entire time?  Will there be any interruption to data plane traffic?

**Comment by adleong** [comment]
Just to make sure I understand: when installing with Helm, is the expectation is that namespace will already exist because either

a) the user created the linkerd namespace ahead of time or
b) the user passed the --create-namepsace flag to helm install

is that right?

File: cli/cmd/install.go
Comment (Pothulapati) [inline cli/cmd/install.go:430]: Hmm, This is removed because. Now we don't need to find if `namespace` has been overriden through helm values, etc, and there's now only  a single way to override namespace which is through `--namespace` flag propogated into `controlPlaneNamespace`, right?

**Review by Pothulapati — COMMENTED** [review]
Changes LGTM but I would love to know more about the upgrade procedure. Especially if we can figure out a no down-time way of upgrading existing linkerd Helm installs.

From what I understand, The real pain-point with Helm is its secret which is being stored into the default namespace. (irrespective of which namespace the control-plane is being installed). Can we define some manual steps through which we can move that secret into the namespace in which `linkerd` is installed into, and from there on a normal upgrade would be sufficient? I guess we would be left with some stale namespace values when we are moving the secret but this should still work because we would be passing the right `--namespace` flag with the upgrade which should correctly set all the values as expected🤔

File: test/integration/install_test.go
Comment (Pothulapati) [inline test/integration/install_test.go:512]: These are `vizArgs` right? and we aren't removing that in this PR right? :thinking:

**Comment by alpeb** [comment]
Thanks gentlemen for the reviews :+1: 

> What will the upgrade procedure be for Helm users? I take it they have to do a uninstall and then reinstall of the helm chart? Assuming they use the same issuer certificates, can they keep the data plane running the entire time? Will there be any interruption to data plane traffic?

Yeah, they'll have to uninstall the chart and install the new chart again. The proxies in the data plane will log errors while the control plane gets reinstalled, but from my testing, traffic continues to flow between data plane pods. And after the control plane is available again, the proxies stop logging errors.

>  when installing with Helm, is the expectation is that namespace will already exist because either
> 
> a) the user created the linkerd namespace ahead of time or
> b) the user passed the --create-namepsace flag to helm install
> 
> is that right?

That's correct.

> Can we define some manual steps through which we can move that secret into the namespace in which `linkerd` is installed into, and from there on a normal upgrade would be sufficient?

Sounds tricky, not sure if a simple move will suffice. Also note there's also an upcoming PR that will split the core Helm chart into two (#6620), so we'll end up with totally different charts, which means we'll have to uninstall the old one anyways.

File: test/integration/install_test.go
Comment (alpeb) [reply test/integration/install_test.go:512]: Correct, that's addressed in #6638

File: cli/cmd/install.go
Comment (alpeb) [reply cli/cmd/install.go:430]: That's correct.

**Comment by alpeb** [comment]
Note: please don't merge this just yet, as I want to have all the Helm charts changes (including #6620) to be included in the same edge :wink:

**Comment by Pothulapati** [comment]
> Sounds tricky, not sure if a simple move will suffice. Also note there's also an upcoming PR that will split the core Helm chart into two (#6620), so we'll end up with totally different charts, which means we'll have to uninstall the old one anyways.

Hmm, Makes sense!

**Review by mateiidavid — COMMENTED** [review]
Tested this out and it works really well. Confirmed the changes by installing with helm (`helm install --create-namespace --namespace linkerd`) and also with (`helm install --generate-namespace --namespace dev`). Both worked well to install the control plane.

Also looked in the configmap to see if we are missing the namespace field (since that's no longer persisted in the cm), and also had a look at the injector manifest to double check the flag is in. Can confirm everything worked well and control plane was up and running.

The only thing here is that when we create a namespace that doesn't have the default name (e.g `dev`), checks and injections fail unless we explicitly pass in `--linkerd-namespace`. I guess this is by design though since we have historically supported installing in a separate namespace for a while -- I've never personally done it in tests until now so thought I'd mention it.

I'll hold off on approval until the rest of the changes are merged in, will want to re-test it everything then but so far everything seems to be 👌🏻 👌🏻 .

**Comment by alpeb** [comment]
I've just merged #6638 here and updated this PR's description accordingly.

**Comment by alpeb** [comment]
I have merged here the dependent PRs that have been approved, and copied relevant info into the description above. This is still not to be merged until #6620 gets addressed.

**Comment by alpeb** [comment]
This is ready for review again, together with #6691. Besides the original issue, more context can be obtained from [my presentation in the community meeting](https://www.youtube.com/watch?v=vujvDltxmhg&t=2070s). I also posted an update to the [helm docs](https://github.com/linkerd/website/blob/e3477b9e4a6d8906c1d69a48b8a720994fa35840/linkerd.io/content/2.12/tasks/install-helm.md) and [migration instructions](https://github.com/linkerd/website/blob/e3477b9e4a6d8906c1d69a48b8a720994fa35840/linkerd.io/content/2.12/tasks/upgrade.md#upgrade-notice-stable-2120) in the linkerd docs.

**Review by adleong — APPROVED** [review]
:+1:

What is the upgrade procedure for Helm users upgrading from a previous version?  What does it mean, specifically, that the chart in 2.11 is backwards-incompatible?

**Comment by alpeb** [comment]
> What is the upgrade procedure for Helm users upgrading from a previous version? What does it mean, specifically, that the chart in 2.11 is backwards-incompatible?

People won't be able to perform a helm upgrade to update to the 2.12 charts. As described in the [upgrade notes](https://github.com/linkerd/website/blob/e3477b9e4a6d8906c1d69a48b8a720994fa35840/linkerd.io/content/2.12/tasks/upgrade.md#upgrade-notice-stable-2120) to migrate they need to first make sure they have all their values.yaml customizations, uninstall the old chart, install the two new ones making sure they pass the appropriate values to each chart, and rollout their control planes as usual. This migration shouldn't incur in any dataplane downtime.

**Review by kleimkuhler — DISMISSED** [review]
Thanks for piping all this through @alpeb it looks good to me.

File: charts/linkerd2/values.yaml
Comment (mateiidavid) [inline charts/linkerd2/values.yaml:372]: Perhaps a bit besides the point but this was commented before, do we still want to have it commented by default?

File: viz/charts/linkerd-viz/templates/tap.yaml
Comment (mateiidavid) [inline viz/charts/linkerd-viz/templates/tap.yaml:36]: If tap will get injected through the injector, do we need to be explicit about `proxy-await`? It should be the default injector behaviour to add await now post 2.10.1 so we might be able to remove it?

File: viz/charts/linkerd-viz/templates/prometheus.yaml
Comment (mateiidavid) [inline viz/charts/linkerd-viz/templates/prometheus.yaml:202]: Same as with tap, I think we can maybe remove the await annotation if it's admitted by the injector? (which afaik should add the await postStart hook by default)

**Review by mateiidavid — APPROVED** [review]
Everything works great! No blockers from my side, thanks for all of the work @alpeb!

I left two comments that can be addressed at any point; the first is about `spValidatorResources` being uncommented and the second is about using `linkerd-await` in viz. It's not really a problem to be explicit from my point of view, so I'm fine with leaving the await annotations in there too, just wanted to point out they may not be necessary 🤔 

🚢 📦

**Comment by kleimkuhler** [comment]
@mateiidavid It's possible to disable the `proxyAwait` feature cluster-wide with the `--set proxy.await=false` flag. When this is the case, it's usually because a user wants their application containers to be first in the list due to other tools that make this assumption; we still want Linkerd components to wait for their proxy to be ready though.

Hardcoding the annotation like we in the places that you've commented ensures that _at least_ all Linkerd components will still wait for their proxy container to be ready.

**Comment by mateiidavid** [comment]
@kleimkuhler ah thanks, yeah that makes a lot of sense. I failed to consider this might be turned off. Disregard what I said then this is definitely good to 🚢 from my side!

File: charts/linkerd2/values.yaml
Comment (alpeb) [reply charts/linkerd2/values.yaml:285]: I uncommented it so that the description is picked up in the chart's README.md file :wink: . Doing so doesn't impact the templates I think (no empty resources appear or anything like that).

**Comment by alpeb** [comment]
I merged #6691 here and updated the description accordingly.

**Comment by alpeb** [comment]
After some feedback we decided to rename `linkerd-base` to `linkerd-crds`, which now only holds the CRDs.

**Review by kleimkuhler — APPROVED** [review]

**Pull request merged**
